README.md
# InkWellSplash

[![Pub Package](https://img.shields.io/pub/v/inkwell_splash.svg)](https://pub.dev/packages/inkwell_splash)
[![Buy Me A Beer](https://img.shields.io/badge/Donate-Buy%20me%20a%20Beer-yellow.svg)](https://www.buymeacoffee.com/manuelrohrauer)
[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/ManuelRohrauer)

Flutter package which starts the splash effect of InkWell instantly if `onDoubleTap` is used.
Additionally maximal time between two taps to recognize a double tap can be adjusted.

### Explanation
The problem with the current InkWell is the delayed start of the splash effect while using onDoubleTap

If you are using `onTap` & `onDoubleTap` side by side with an `InkWell`, then a single tap gets delayed. This is necessary for the recognition of a double tap but also the splash effect is delayed. With this package the splash effect is startet instantly.

All other functions are equal to the standard `InkWell` widget

[Interactive Example](https://dartpad.dev/embed-flutter.html?gh_owner=ManuelRohrauer&gh_repo=inkwell_splash&gh_path=dartpad_example&run=true)
 
### Usage
Add this line to pubspec.yaml
```yaml
dependencies:
  inkwell_splash: ^0.1.0
```
Import package
```dart
import 'package:inkwell_splash/inkwell_splash.dart';
```
Simple example:
```dart
class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Material(
      color: Colors.redAccent,
      child: InkWellSplash(
        splashFactory: InkRipple.splashFactory,
        onTap: () => print("Single Tap"),
        onDoubleTap: () => print("Double Tap"),
        doubleTapTime: Duration(milliseconds: 300),
        child: Container(
          width: 120,
          height: 120,
          child: Text("Welcome to InkWellSplash"),
        ),
      ),
    );
  }
}
```

### Parameters

| Parameter | Description | Type |
|---|---|---|
| `doubleTapTime` | Maximal time between two taps to recognize a double tap | `Duration` |

Additional all parameters of standard InkWell can be used as well

### Note
If another parent or child gesture widget (e.g. InkWell, GestureDetector) uses its own `onDoubleTap` function, the splash of the InkWellSplash is delayed again. This is because, the onTap function of all widgets will then be delayed.



CHANGELOG.md
### [0.1.0] - 18.04.2020
* Removed tap delay if onDoubleTab is not used

### [0.0.1] - 18.04.2020
* Initial Release


lib/inkwell_splash.dart
library inkwell_splash;

import 'package:flutter/material.dart';
import 'dart:async';

class InkWellSplash extends StatelessWidget {
  InkWellSplash({
    Key key,
    this.child,
    this.onTap,
    this.onDoubleTap,
    this.doubleTapTime = const Duration(milliseconds: 300),
    this.onLongPress,
    this.onTapDown,
    this.onTapCancel,
    this.onHighlightChanged,
    this.onHover,
    this.focusColor,
    this.hoverColor,
    this.highlightColor,
    this.splashColor,
    this.splashFactory,
    this.radius,
    this.borderRadius,
    this.customBorder,
    this.enableFeedback = true,
    this.excludeFromSemantics = false,
    this.focusNode,
    this.canRequestFocus = true,
    this.onFocusChange,
    this.autofocus = false,
  })  : assert(enableFeedback != null),
        assert(excludeFromSemantics != null),
        assert(autofocus != null),
        assert(canRequestFocus != null),
        super(key: key);

  final Widget child;
  final GestureTapCallback onTap;
  final GestureTapCallback onDoubleTap;
  final Duration doubleTapTime;
  final GestureLongPressCallback onLongPress;
  final GestureTapDownCallback onTapDown;
  final GestureTapCancelCallback onTapCancel;
  final ValueChanged<bool> onHighlightChanged;
  final ValueChanged<bool> onHover;
  final Color focusColor;
  final Color hoverColor;
  final Color highlightColor;
  final Color splashColor;
  final InteractiveInkFeatureFactory splashFactory;
  final double radius;
  final BorderRadius borderRadius;
  final ShapeBorder customBorder;
  final bool enableFeedback;
  final bool excludeFromSemantics;
  final FocusNode focusNode;
  final bool canRequestFocus;
  final ValueChanged<bool> onFocusChange;
  final bool autofocus;

  Timer doubleTapTimer;
  bool isPressed = false;
  bool isSingleTap = false;
  bool isDoubleTap = false;

  void _doubleTapTimerElapsed() {
    if (isPressed) {
      isSingleTap = true;
    } else {
      if(this.onTap != null) this.onTap();
    }
  }

  void _onTap() {
    isPressed = false;
    if (isSingleTap) {
      isSingleTap = false;
      if(this.onTap != null) this.onTap();           // call user onTap function
    }
    if (isDoubleTap) {
      isDoubleTap = false;
      if(this.onDoubleTap != null) this.onDoubleTap();
    }
  }

  void _onTapDown(TapDownDetails details) {
    isPressed = true;
    if (doubleTapTimer != null && doubleTapTimer.isActive) {
      isDoubleTap = true;
      doubleTapTimer.cancel();
    } else {
      doubleTapTimer = Timer(doubleTapTime, _doubleTapTimerElapsed);
    }
    if(this.onTapDown != null) this.onTapDown(details);
  }

  void _onTapCancel() {
    isPressed = isSingleTap = isDoubleTap = false;
    if (doubleTapTimer != null && doubleTapTimer.isActive) {
      doubleTapTimer.cancel();
    }
    if(this.onTapCancel != null) this.onTapCancel();
  }

  @override
  Widget build(BuildContext context) {
    return InkWell(
      key: key,
      child: child,
      onTap: (onDoubleTap != null) ? _onTap : onTap,    // if onDoubleTap is not used from user, then route further to onTap
      onLongPress: onLongPress,
      onTapDown: (onDoubleTap != null) ? _onTapDown : onTapDown,
      onTapCancel: (onDoubleTap != null) ? _onTapCancel : onTapCancel,
      onHighlightChanged: onHighlightChanged,
      onHover: onHover,
      focusColor: focusColor,
      hoverColor: hoverColor,
      highlightColor: highlightColor,
      splashColor: splashColor,
      splashFactory: splashFactory,
      radius: radius,
      borderRadius: borderRadius,
      customBorder: customBorder,
      enableFeedback: enableFeedback,
      excludeFromSemantics: excludeFromSemantics,
      focusNode: focusNode,
      canRequestFocus: canRequestFocus,
      onFocusChange: onFocusChange,
      autofocus: autofocus,
    );
  }
}


dartpad_example/dartpad_inkwell_splash.dart
import 'package:flutter/material.dart';
import 'dart:async';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: MyWidget(),
        ),
      ),
    );
  }
}

class MyWidget extends StatefulWidget {
  @override
  _MyWidgetState createState() => _MyWidgetState();
}

class _MyWidgetState extends State<MyWidget> {
  int inkWellTaps = 0;
  int inkWellDoubleTaps = 0;
  int inkWellSplashTaps = 0;
  int inkWellSplashDoubleTaps = 0;

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.spaceEvenly,
      children: [
        Material(
          color: Colors.redAccent,
          child: InkWell(
            splashFactory: InkRipple.splashFactory,
            onTap: () => setState(() => inkWellTaps++),
            onDoubleTap: () => setState(() => inkWellDoubleTaps++),
            child: Container(
              alignment: Alignment.center,
              width: 120,
              height: 120,
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                //crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  Text('Standard InkWell', style: TextStyle(fontWeight: FontWeight.bold)),
                  SizedBox(height: 20),
                  Text('Taps: $inkWellTaps'),
                  Text('Double Taps: $inkWellDoubleTaps'),
                ],
              ),
            ),
          ),
        ),
        Material(
          color: Colors.redAccent,
          child: InkWellSplash(
            splashFactory: InkRipple.splashFactory,
            onTap: () => setState(() => inkWellSplashTaps++),
            onDoubleTap: () => setState(() => inkWellSplashDoubleTaps++),
            doubleTapTime: Duration(milliseconds: 300),
            child: Container(
              width: 120,
              height: 120,
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                //crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  Text('InkWellSplash', style: TextStyle(fontWeight: FontWeight.bold)),
                  SizedBox(height: 20),
                  Text('Taps: $inkWellSplashTaps'),
                  Text('Double Taps: $inkWellSplashDoubleTaps'),
                ],
              ),
            ),
          ),
        ),
      ],
    );
  }
}

// This is the actual package file, this is here to use the interactive example with dartpad correctly
class InkWellSplash extends StatelessWidget {
  InkWellSplash({
    Key key,
    this.child,
    this.onTap,
    this.onDoubleTap,
    this.doubleTapTime = const Duration(milliseconds: 300),
    this.onLongPress,
    this.onTapDown,
    this.onTapCancel,
    this.onHighlightChanged,
    this.onHover,
    this.focusColor,
    this.hoverColor,
    this.highlightColor,
    this.splashColor,
    this.splashFactory,
    this.radius,
    this.borderRadius,
    this.customBorder,
    this.enableFeedback = true,
    this.excludeFromSemantics = false,
    this.focusNode,
    this.canRequestFocus = true,
    this.onFocusChange,
    this.autofocus = false,
  })  : assert(enableFeedback != null),
        assert(excludeFromSemantics != null),
        assert(autofocus != null),
        assert(canRequestFocus != null),
        super(key: key);

  final Widget child;
  final GestureTapCallback onTap;
  final GestureTapCallback onDoubleTap;
  final Duration doubleTapTime;
  final GestureLongPressCallback onLongPress;
  final GestureTapDownCallback onTapDown;
  final GestureTapCancelCallback onTapCancel;
  final ValueChanged<bool> onHighlightChanged;
  final ValueChanged<bool> onHover;
  final Color focusColor;
  final Color hoverColor;
  final Color highlightColor;
  final Color splashColor;
  final InteractiveInkFeatureFactory splashFactory;
  final double radius;
  final BorderRadius borderRadius;
  final ShapeBorder customBorder;
  final bool enableFeedback;
  final bool excludeFromSemantics;
  final FocusNode focusNode;
  final bool canRequestFocus;
  final ValueChanged<bool> onFocusChange;
  final bool autofocus;

  Timer doubleTapTimer;
  bool isPressed = false;
  bool isSingleTap = false;
  bool isDoubleTap = false;

  void _doubleTapTimerElapsed() {
    if (isPressed) {
      isSingleTap = true;
    } else {
      if(this.onTap != null) this.onTap();
    }
  }

  void _onTap() {
    isPressed = false;
    if (isSingleTap) {
      isSingleTap = false;
      if(this.onTap != null) this.onTap();           // call user onTap function
    }
    if (isDoubleTap) {
      isDoubleTap = false;
      if(this.onDoubleTap != null) this.onDoubleTap();
    }
  }

  void _onTapDown(TapDownDetails details) {
    isPressed = true;
    if (doubleTapTimer != null && doubleTapTimer.isActive) {
      isDoubleTap = true;
      doubleTapTimer.cancel();
    } else {
      doubleTapTimer = Timer(doubleTapTime, _doubleTapTimerElapsed);
    }
    if(this.onTapDown != null) this.onTapDown(details);
  }

  void _onTapCancel() {
    isPressed = isSingleTap = isDoubleTap = false;
    doubleTapTimer.cancel();
    if(this.onTapCancel != null) this.onTapCancel();
  }

  @override
  Widget build(BuildContext context) {
    return InkWell(
      key: key,
      child: child,
      onTap: (onDoubleTap != null) ? _onTap : onTap,    // if onDoubleTap is not used from user, then route further to onTap
      onLongPress: onLongPress,
      onTapDown: (onDoubleTap != null) ? _onTapDown : onTapDown,
      onTapCancel: (onDoubleTap != null) ? _onTapCancel : onTapCancel,
      onHighlightChanged: onHighlightChanged,
      onHover: onHover,
      focusColor: focusColor,
      hoverColor: hoverColor,
      highlightColor: highlightColor,
      splashColor: splashColor,
      splashFactory: splashFactory,
      radius: radius,
      borderRadius: borderRadius,
      customBorder: customBorder,
      enableFeedback: enableFeedback,
      excludeFromSemantics: excludeFromSemantics,
      focusNode: focusNode,
      canRequestFocus: canRequestFocus,
      onFocusChange: onFocusChange,
      autofocus: autofocus,
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:inkwell_splash/inkwell_splash.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: MyWidget(),
        ),
      ),
    );
  }
}

class MyWidget extends StatefulWidget {
  @override
  _MyWidgetState createState() => _MyWidgetState();
}

class _MyWidgetState extends State<MyWidget> {
  int inkWellTaps = 0;
  int inkWellDoubleTaps = 0;
  int inkWellSplashTaps = 0;
  int inkWellSplashDoubleTaps = 0;

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.spaceEvenly,
      children: [
        Material(
          color: Colors.redAccent,
          child: InkWell(
            splashFactory: InkRipple.splashFactory,
            onTap: () => setState(() => inkWellTaps++),
            onDoubleTap: () => setState(() => inkWellDoubleTaps++),
            child: Container(
              alignment: Alignment.center,
              width: 120,
              height: 120,
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  Text('Standard InkWell', style: TextStyle(fontWeight: FontWeight.bold)),
                  SizedBox(height: 20),
                  Text('Taps: $inkWellTaps'),
                  Text('Double Taps: $inkWellDoubleTaps'),
                ],
              ),
            ),
          ),
        ),
        Material(
          color: Colors.redAccent,
          child: InkWellSplash(
            splashFactory: InkRipple.splashFactory,
            onTap: () => setState(() => inkWellSplashTaps++),
            onDoubleTap: () => setState(() => inkWellSplashDoubleTaps++),
            doubleTapTime: Duration(milliseconds: 300),
            child: Container(
              width: 120,
              height: 120,
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  Text('InkWellSplash', style: TextStyle(fontWeight: FontWeight.bold)),
                  SizedBox(height: 20),
                  Text('Taps: $inkWellSplashTaps'),
                  Text('Double Taps: $inkWellSplashDoubleTaps'),
                ],
              ),
            ),
          ),
        ),
      ],
    );
  }
}
